### PR TITLE
Add resizable behavior for kivy widgets

### PR DESCRIPTION
**Pull request by Bakterija**
Hello. I was thinking that this kind of resizable behavior would be useful to many kivy coders and made a kivy style behavior from it that i would like to merge now.
Can't remember where from i got the icon, but it's this one http://www.creativehunt.co/icons/resize-icon/.
It says that license is public domain, not sure if/where i have to add it in kivy.

### CODE REVIEW

**Comment by KeyWeeUsr** [comment]
Only with these changes it won't work. You need to edit the `__init__.py` of `behaviors` too i.e. add to `__all__` the name of a behavior and don't forget to import it.

File: kivy/uix/behaviors/resize.py [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+
+    def on_leave_resizable(self):
+        Window.show_cursor = True
+        self.cursor.hidden = True
+
+    def on_mouse_move(self, pos):
+        if self.hovering:
+            if self.cursor:
+                self.cursor.on_mouse_move(pos)
+            if not self.resizing:
```
Comment (KeyWeeUsr) [inline kivy/uix/behaviors/resize.py]: this one could be put into upper if (and you'll spare 4 spaces) + it'll look cleaner

File: kivy/uix/behaviors/resize.py [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+        Window.show_cursor = False
+        self.cursor.hidden = False
+
+    def on_leave_resizable(self):
+        Window.show_cursor = True
+        self.cursor.hidden = True
+
+    def on_mouse_move(self, pos):
+        if self.hovering:
+            if self.cursor:
```
Comment (KeyWeeUsr) [inline kivy/uix/behaviors/resize.py]: this could be in the `if` below

File: kivy/uix/behaviors/resize.py [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+                self.resizing = True
+                Window.show_cursor = False
+                return True
+            else:
+                return super(ResizableBehavior, self).on_touch_down(touch)
+        else:
+            return super(ResizableBehavior, self).on_touch_down(touch)
+
+    def on_touch_move(self, touch):
+        if self.resizing:
```
Comment (KeyWeeUsr) [inline kivy/uix/behaviors/resize.py]: `if not self.resizing: return super(...)` and another block down

File: kivy/uix/behaviors/resize.py [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+                            self.height = self.rborder * 3
+            elif self.resizing_up:
+                if touch.pos[1] > self.pos[1] + (self.rborder * 3):
+                    self.height = touch.pos[1] - self.pos[1]
+            return True
+        else:
+            return super(ResizableBehavior, self).on_touch_move(touch)
+
+    def on_touch_up(self, touch):
+        if self.resizing:
```
Comment (KeyWeeUsr) [inline kivy/uix/behaviors/resize.py]: same as `on_touch_move`

File: kivy/uix/behaviors/resize.py [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+    resizing = BooleanProperty(False)
+    '''State of widget resizing.
+    It is switched to True when a resize border is touched and back to False when it is released.
+    It manages resizing when touch is moved.
+
+    :attr:`resizing` is a :class:`~kivy.properties.BooleanProperty` and
+    defaults to False.
+    '''
+
+    dont_move = BooleanProperty(False)
```
Comment (KeyWeeUsr) [inline kivy/uix/behaviors/resize.py]: don't name properties with "`dont`", make them `do_*`. The `if` statement will look cleaner (found below)

File: kivy/uix/behaviors/resize.py [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+        def build(self):
+            return Sample()
+
+
+    SampleApp().run()
+
+See :class:`~kivy.uix.behaviors.ResizableBehavior` for details.
+'''
+
+from __future__ import print_function
```
Comment (Bakterija) [inline kivy/uix/behaviors/resize.py]: PEP8 check was done with pycharm, should be fine.

**Comment by KeyWeeUsr** [comment]
@Bakterija set then the options properly 😋 
```
resize.py:7:80: E501 line too long (86 > 79 characters)
resize.py:8:80: E501 line too long (84 > 79 characters)
resize.py:34:80: E501 line too long (85 > 79 characters)
resize.py:36:80: E501 line too long (88 > 79 characters)
resize.py:46:80: E501 line too long (96 > 79 characters)
resize.py:82:80: E501 line too long (98 > 79 characters)
resize.py:100:80: E501 line too long (82 > 79 characters)
resize.py:103:80: E501 line too long (84 > 79 characters)
resize.py:107:80: E501 line too long (81 > 79 characters)
resize.py:119:80: E501 line too long (80 > 79 characters)
resize.py:137:80: E501 line too long (104 > 79 characters)
resize.py:138:80: E501 line too long (90 > 79 characters)
resize.py:139:80: E501 line too long (99 > 79 characters)
resize.py:156:80: E501 line too long (98 > 79 characters)
resize.py:158:80: E501 line too long (81 > 79 characters)
resize.py:199:80: E501 line too long (97 > 79 characters)
resize.py:200:80: E501 line too long (95 > 79 characters)
resize.py:208:80: E501 line too long (98 > 79 characters)
resize.py:209:80: E501 line too long (96 > 79 characters)
resize.py:217:80: E501 line too long (98 > 79 characters)
resize.py:218:80: E501 line too long (96 > 79 characters)
resize.py:226:80: E501 line too long (98 > 79 characters)
resize.py:227:80: E501 line too long (96 > 79 characters)
resize.py:236:80: E501 line too long (97 > 79 characters)
resize.py:245:80: E501 line too long (87 > 79 characters)
resize.py:247:80: E501 line too long (96 > 79 characters)
resize.py:260:80: E501 line too long (97 > 79 characters)
resize.py:324:80: E501 line too long (114 > 79 characters)
resize.py:325:80: E501 line too long (95 > 79 characters)
resize.py:332:80: E501 line too long (100 > 79 characters)
resize.py:349:80: E501 line too long (88 > 79 characters)
resize.py:352:80: E501 line too long (84 > 79 characters)
resize.py:358:80: E501 line too long (88 > 79 characters)
resize.py:361:80: E501 line too long (85 > 79 characters)
```

File: kivy/uix/behaviors/resize.py (hunk 5 above)
Comment (Bakterija) [inline kivy/uix/behaviors/resize.py]: changed variable to "can_resize_move" and docs to explain that this controls if the widget is allowed to be moved on resize. is that good?

File: kivy/uix/behaviors/resize.py (hunk 3 above)
Comment (Bakterija) [inline kivy/uix/behaviors/resize.py]: I'am not sure what you mean. I have to go to super function to make it compatable with button behavior and similiar modifications.

File: kivy/uix/behaviors/resize.py (hunk 5 above)
Comment (KeyWeeUsr) [inline kivy/uix/behaviors/resize.py]: Seems better, but why not rather `resize_move`? I don't want to sound annoying or something, but properties in `my_long_property_style` or `dont_do_this` aren't nice and may be confusing when reading the code.

File: kivy/uix/behaviors/resize.py (hunk 3 above)
Comment (KeyWeeUsr) [inline kivy/uix/behaviors/resize.py]: def on_touch_move(self, touch):
        if not self.resizing:
            return super(...)
        <the rest of code that would be in "if self.resizing">

**Comment by Bakterija** [comment]
I managed to shorten all lines to below 80 characters

**Comment by KeyWeeUsr** [comment]
You see, the thing I meant by the moving `else` block to the top is because it **_returns_**. If a block returns, the rest of the code isn't executed, which means, the `else` block now doesn't need to exist. It can be just a casual block of code without if-else:

    def method(self, something):
        if something:
            print(1)
            print(2)
        else:
            return 0

can be simply this:

    def method(self, something):
        if something:
            return 0

        print(1)  # no indent wasting, no staircase in code
        print(2)

File: kivy/uix/behaviors/resize.py (hunk 5 above)
Comment (Bakterija) [inline kivy/uix/behaviors/resize.py]: I aggree, short variable names are good, but I also like to describe what I am doing.I am not sure what to name it.
But i would rename it to resize_move, if another dev aggrees that it's a good name.

File: kivy/uix/behaviors/resize.py (hunk 3 above)
Comment (Bakterija) [inline kivy/uix/behaviors/resize.py]: I put those at beginning of function now.

File: kivy/uix/behaviors/resize.py (hunk 2 above)
Comment (Bakterija) [inline kivy/uix/behaviors/resize.py]: Ok. Merged line 277 with line 278

**Comment by Bakterija** [comment]
You are right about the return. I was just thinking that I might have to add/change some code there later.
Forgot and will remove the redundant "else:" blocks right now

File: kivy/uix/behaviors/resize.py [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+    hovering_resizable = BooleanProperty(False)
+    '''State of mouse hover.
+    It is switched to True when mouse is inside the widgets resize border
+    and False when it isn't.
+
+    :attr:`hovering_resizable` is a :class:`~kivy.properties.BooleanProperty`
+    and defaults to False.
+    '''
+
+    rborder = NumericProperty(cm(0.5))
```
Comment (rnixx) [inline kivy/uix/behaviors/resize.py]: also I'd not use Centimeters as metric. ``dp`` would be my choice.

Comment (Bakterija) [inline kivy/uix/behaviors/resize.py]: Good ideas , I changed it.

File: kivy/uix/behaviors/resize.py (hunk 6 above)
Comment (Bakterija) [inline kivy/uix/behaviors/resize.py]: Removed futures

File: kivy/uix/behaviors/resize.py [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+    def on_enter_resizable(self):
+        Window.show_cursor = False
+        self.cursor.hidden = False
+
+    def on_leave_resizable(self):
+        Window.show_cursor = True
+        self.cursor.hidden = True
+
+    def on_mouse_move(self, pos):
+        if self.hovering and self.cursor:
```
Comment (KeyWeeUsr) [inline kivy/uix/behaviors/resize.py]: I still think this:
```
if self.hovering and not self.resizing:
    if self.cursor:
        do_something()
    <the rest>
```
is prettier than the current alternative, the less "stairs" in code, the more readable it is.

Comment (Bakterija) [inline kivy/uix/behaviors/resize.py]: This change will make the cursor stay in one place while resizing.
I made it like that to limit cursor.pos updates to only when it's necessary.

Comment (KeyWeeUsr) [inline kivy/uix/behaviors/resize.py]: Oh, ok then, sorry. Now it's only pep8 🍡

**Comment by Bakterija** [comment]
Added min and max size properties. Now only offset (for resizing outside widgets) remaining, and maybe some day we will have system cursors.

**Comment by FeralBytes** [comment]
Just subscribing to this PR so I might help with review; I am not a dev for kivy; just a user.
@tshirtman when able please post the problem areas that you foresaw in your review.

**Comment by dessant** [comment]
As discussed on irc, there were some reservations regarding this, though they are not yet well formed. 😋
Meanwhile the feature was added to garden and it may be merged back in the future, when its scope and effects become a bit more clear.

https://github.com/kivy-garden/garden.resizable_behavior

**Pull request closed without merging**